src/vdaf/mastic.rs

```diff
@@ -0,0 +1,547 @@
+    flp::{
+        szk::{Szk, SzkProofShare},
+        Type,
+    },
+    vdaf::{
+        poplar1::Poplar1AggregationParam,
+        xof::{Seed, Xof},
+        AggregateShare, Client, OutputShare, Vdaf, VdafError,
+    },
+    vidpf::{
+        Vidpf, VidpfError, VidpfInput, VidpfKey, VidpfPublicShare, VidpfServerId, VidpfWeight,
+    },
+};
+
+use std::fmt::Debug;
+use std::io::{Cursor, Read};
+use std::ops::{BitAnd, Not};
+use subtle::{Choice, ConstantTimeEq};
+
+/// The main struct implementing the Mastic VDAF.
+/// Composed of a shared zero knowledge proof system and a verifiable incremental
+/// distributed point function.
+#[derive(Clone, Debug)]
+pub struct Mastic<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    algorithm_id: u32,
+    szk: Szk<T, P, SEED_SIZE>,
+    pub(crate) vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+    /// The length of the private attribute associated with any input.
+    pub(crate) bits: usize,
+}
+
+impl<T, P, const SEED_SIZE: usize> Mastic<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Creates a new instance of Mastic, with a specific attribute length and weight type.
+    pub fn new(
+        algorithm_id: u32,
+        szk: Szk<T, P, SEED_SIZE>,
+        vidpf: Vidpf<VidpfWeight<T::Field>, 16>,
+        bits: usize,
+    ) -> Self {
+        Self {
+            algorithm_id,
+            szk,
+            vidpf,
+            bits,
+        }
+    }
+}
+
+/// Mastic aggregation parameter.
+///
+/// This includes the VIDPF tree level under evaluation and a set of prefixes to evaluate at that level.
+#[derive(Clone, Debug)]
+pub struct MasticAggregationParam {
+    /// aggregation parameter inherited from [`Poplar1`]: contains the level (attribute length) and a vector of attribute prefixes (IdpfInputs)
+    level_and_prefixes: Poplar1AggregationParam,
+    /// Flag indicating whether the VIDPF weight needs to be validated using SZK.
+    /// This flag must be set the first time any report is aggregated; however this may happen at any level of the tree.
+    require_check_flag: bool,
+}
+
+impl Encode for MasticAggregationParam {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        self.level_and_prefixes.encode(bytes)?;
+        let require_check = if self.require_check_flag { 1u8 } else { 0u8 };
+        require_check.encode(bytes)?;
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        Some(self.level_and_prefixes.encoded_len()? + 1usize)
+    }
+}
+
+impl Decode for MasticAggregationParam {
+    fn decode(bytes: &mut Cursor<&[u8]>) -> Result<Self, CodecError> {
+        let level_and_prefixes = Poplar1AggregationParam::decode(bytes)?;
+        let require_check = u8::decode(bytes)?;
+        let require_check_flag = require_check != 0;
+        Ok(Self {
+            level_and_prefixes,
+            require_check_flag,
+        })
+    }
+}
+
+/// Mastic public share
```

Comment:
nit: Be consistent about adding a period after the first paragraph (here and below)
```suggestion
/// Mastic public share.
```